keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,358 @@
+from within the cluster or not.
+
+### ipvs - accepting traffic after first packet, after being accepted by kubernetes rules
+
+The [rule here currently](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1635-L1654)
+looks as follows
+
+```go
+// The following two rules ensure the traffic after the initial packet
+// accepted by the "kubernetes forwarding rules" rule above will be
+// accepted, to be as specific as possible the traffic must be sourced
+// or destined to the clusterCIDR (to/from a pod).
+writeLine(proxier.filterRules,
+	"-A", string(KubeForwardChain),
+	"-s", proxier.clusterCIDR,
+	"-m", "comment", "--comment", `"kubernetes forwarding conntrack pod source rule"`,
+	"-m", "conntrack",
+	"--ctstate", "RELATED,ESTABLISHED",
+	"-j", "ACCEPT",
+)
+writeLine(proxier.filterRules,
+	"-A", string(KubeForwardChain),
+	"-m", "comment", "--comment", `"kubernetes forwarding conntrack pod destination rule"`,
+	"-d", proxier.clusterCIDR,
+	"-m", "conntrack",
+	"--ctstate", "RELATED,ESTABLISHED",
+	"-j", "ACCEPT",
+)
+```
+Again, applying similar logic to the last rule for iptables, the proposal here is to simplify this drop reference to
+the proxy.clusterCIDR and just match on the connection state.
+
+### Risks and Mitigations
+
+The biggest risk we have is that we are expanding the scope of the last rule to potentially include non-kubernetes
+traffic. This is considered mostly safe as it does not break any of the intended drop behavior. Plus once the initial
+connection has been accepted, assuming nodes are used for kubernetes workloads, it's highly unlikely that we would
+need to not accept it later.
+
+## Design Details
+
+The idea of ‘determine cluster originated traffic’ would be captured in a new go interface type within kube-proxy,
+with different implementations of the interface. The kube-proxy implementation itself would just call method on
+interface to get the match criteria to write in the rule.
+
+The new behavior can be opted-in using the following flags as an alternative to `--cluster-cidr string` flag.
+
+```
+--detect-local-with-node-cidr string
+
+  kube-proxy considers traffic as local if originating from within this cidr list. string argument 
+  can be an empty string. If empty (""), then kube-proxy will use the node.spec.podCIDR
+  attribute to identify the list of pod cidr. Can be a comma separated list of CIDR in a.b.c.d/x format
+  in case manually specified
+
+--detect-local-with-pod-interface string
+
+  kube-proxy considers traffic as local if originating from an interface which matches one of given
+  prefixes. string argument is a comma separated list of interface prefix names, without the ending '+'.
+
+--detect-local-with-bridge
+
+  kube-proxy considers traffic as local if originating from a bridge within the node.
+```
+
+Only one of `--cluster-cidr`, `--detect-local-with-node-cidr`, `--detect-local-with-pod-interface` or
```

Comment:
I was thinking something like:

`--detect-local={cluster-cidr | node-cidr | pod-interface-prefix | bridge}` defaulting to cluster-cidr for compat

* if cluster-cidr, look at the --cluster-cidr flag
* if node-cidr, look at the node PodCIDR field
* if interface-prefix, look at the --pod-interface-prefix flag
* if bridge, use physdev

This could extend to "cidr", with a list of specific CIDRs, "mark" with a specific mark value/mask, etc.  This seems less ambiguous to me, but I am not wedded to it, either.

Reply:
Thanks for this. I really like this because if we default --detect-local=cluster-cidr, then the current behavior becomes the automatic default when not using any new flags. The only change I would make to this would be to adding an optional --pod-cidr flag to go along with the node-cidr choice to handle cases where node CIDR(s) are not tracked as part of the node.podCIDR field.

Will update the kep with the new flag format.

Reply:
Not to bikeshed, but `--detect-reachable`, since this isn't really about locality, but reachability?

Reply:
I put in my perspective on the following comment. I think "--detect-local" does capture the intent better? I don't think it's about reachability.

Reply:
Have updated this implementation section with the new flags. Can this be resolved?